Docs/documentation 2.0 (#158)
sse-secure-systems/connaisseur: docs/validators/notaryv1.md

    key: |
      -----BEGIN PUBLIC KEY-----
      MFkwEwYHKoZIzj0CAQYIKoZIzj0DAQcDQgAEvtc/qpHtx7iUUj+rRHR99a8mnGni
      qiGkmUb9YpWWTS4YwlvwdmMDiGzcsHiDOYz6f88u2hCRF5GUCvyiZAKrsA==
      -----END PUBLIC KEY-----

policy:
- pattern: "docker.io/securesystemsengineering/connaisseur:*"
  validator: docker_essentials
  with:
    key: sse
    delegations:
    - belitzphilipp
    - starkteetje
```

## Delegations

Notary (V1) offers the functionality to delegate trust. To better understand this feature, it's best to have a basic understanding of the TUF key hierarchy. If you are more interested in this topic, please read their [documentation](https://theupdateframework.github.io/specification/latest/#roles-and-pki).

When creating the signatures of your docker images earlier.

## Additional notes

### Getting the Root Key

For verifying the signatures of images, a public root key is needed, but from where do you get this, especially for public images whose signatures you didn't create? For this we created the *get_root_key* utility. You can use it by building the docker image with `docker build -t get-root-key -f docker/Dockerfile.getRoot .` and the running the image:

```bash

> $ docker run --rm notaryrootkey -i securesystemsengineering/testimage

Reviewer: ```suggestion
$ docker run --rm get-root-key -i securesystemsengineering/testimage
```

# Notary (V1) / DCT

> Notary works as an external service holding signatures and trust data based on [TUF](https://theupdateframework.io/) of any kind of data. Official DockerHub uses "Docker Content Trust" (DCT) to sign its docker images and pushes the repository's signature data to its associated Notary server (notary.docker.io). Validating a single docker image requires the public repository root key as well as fetching the repository's trust data from the associated Notary server. Besides of public Notary instaces as e.g. notary.docker.io being ready to use, Notary can as well be run as a private instance. Harbor for  instance comes along with an associated Notary instance. Notary (v1) will be the default validator when signing images using Docker Content Trust (DCT).

Reviewer: ```suggestion
Notary works as an external service holding signatures and trust data based on [TUF](https://theupdateframework.io/) of any kind of data. Official DockerHub uses "Docker Content Trust" (DCT) to sign its docker images and pushes the repository's signature data to its associated Notary server (notary.docker.io). Validating a single docker image requires the public repository root key as well as fetching the repository's trust data from the associated Notary server. Besides of public Notary instances as e.g. notary.docker.io being ready to use, Notary can as well be run as a private instance. Harbor for  instance comes along with an associated Notary instance. Notary (v1) will be the default validator when signing images using Docker Content Trust (DCT).
```

5b8c72934dfc: Pushed 
stable: digest: sha256:dca71257cd2e72840a21f0323234bb2e33fea6d949fa0f21c5102146f583486b size: 527
Signing and pushing trust metadata
You are about to create a new root signing key passphrase. This passphrase
will be used to protect the most sensitive key in your signing system. Please
choose a long, complex passphrase and be careful to keep the password and the
key file itself secure and backed up. It is highly recommended that you use a
password manager to generate the passphrase and keep it safe. There will be no
way to recover this key. You can find the key in your config directory.
Enter passphrase for new root key with ID 5fb3e1e: 
Repeat passphrase for new root key with ID 5fb3e1e: 
Enter passphrase for new repository key with ID 6c2a04c: 
Repeat passphrase for new repository key with ID 6c2a04c: 
Finished initializing "<your-private-repo>/busybox"
```

After doing so, the signature for your image will reside in the public DockerHub Notary (notary.docker.io). Your newly created private keys will be needed whenever you push a new version of your image. They reside in your `~/.docker/trust/private` directory.

> ### Getting the Root Key

Reviewer: Capitalization of headlines is inconsistent

> For verifying the signatures of images, a public [root key](https://theupdateframework.github.io/specification/latest/#root) is needed to verify the signature data with it. But from where do you get this, especially for public images whose signatures you didn't create? For this we created the *get_root_key* utility. You can use it by building the docker image with `docker build -t get-root-key -f docker/Dockerfile.getRoot .` and the running the image:

Reviewer: `For verifying ...is needed to verify` sounds a bit clumsy

> For verifying the signatures of images, a public [root key](https://theupdateframework.github.io/specification/latest/#root) is needed to verify the signature data with it. But from where do you get this, especially for public images whose signatures you didn't create? For this we created the *get_root_key* utility. You can use it by building the docker image with `docker build -t get-root-key -f docker/Dockerfile.getRoot .` and the running the image:

Reviewer: ```suggestion
For verifying the signatures of images, a public [root key](https://theupdateframework.github.io/specification/latest/#root) is needed to verify the signature data with it. But from where do you get this, especially for public images whose signatures you didn't create? For this we created the *get_root_key* utility. You can use it by building the docker image with `docker build -t get-root-key -f docker/Dockerfile.getRoot .` and then running the image:
```

```bash
$ docker run --rm get-root-key -i securesystemsengineering/testimage
KeyID: 76d211ff8d2317d78ee597dbc43888599d691dbfd073b8226512f0e9848f2508
Key: -----BEGIN PUBLIC KEY-----
MFkwEwYHKoZIzj0CAQYIKoZIzj0DAQcDQgAEsx28WV7BsQfnHF1kZmpdCTTLJaWe
d0CA+JOi8H4REuBaWSZ5zPDe468WuOJ6f71E7WFg3CVEVYHuoZt2UYbN/Q==
-----END PUBLIC KEY-----
```

> The `-i` (`--image`) option is required and takes the image, for which you want the public key. There is also the `-s` (`--server`) option, which defines the notary server that should be used and which defaults to `notary.docker.io`.

Reviewer: Previously, you capitalized Notary. Consistency

The public repository root key resides with the signature data in the notary instance, so what the *get_root_key* utility does in the background is just fetching, locating and parsing the public repository root key for the given image.

### Configuration options

The Notary (V1) validator supports the following fields (refer to [basics](../basics.md) for more information on default keys):

| key | default | required | description |
| - | - | - | - |
| `name` | - | :heavy_check_mark: | Unique name of the validator |
| `type` | - | :heavy_check_mark: | This must be set to `notaryv1` |

> | `host` | - | :heavy_check_mark: | URL of the Notary instance, in which the signatures reside. |

Reviewer: This description ends with `.` while prior ones didnt

Delegations will now go one level deeper, meaning they can be used to sign individual image repositories and only need the targets key for rotation purposes, instead of the root key. Also delegation keys are not bound to individual image repositories, so they can be reused multiple times over different image repositories. So in a sense they can be understood as keys for individual signers. 

To create a delegation key run:

```bash
> docker trust key generate <your-name>
Generating key for <your-name>...
Enter passphrase for new <your-name> key with ID 9deed25: 
Repeat passphrase for new <your-name> key with ID 9deed25: 

> Successfully generated and loaded private key. Corresponding public key available: /home/ubuntu/Documents/mytrustdir/<your-name>.pub

Reviewer: ```suggestion
Successfully generated and loaded private key. Corresponding public key available: <current-directory>/<your-name>.pub
```
?

```

> This delegation key now needs to be added as a signer to a respective image repository, like the `busybox` one from above. In doing so, you'll be ask for the targets key.

Reviewer: ```suggestion
This delegation key now needs to be added as a signer to a respective image repository, like the `busybox` one from above. In doing so, you'll be asked for the targets key.
```

```bash
> docker trust signer add --key <your-name>.pub <your-name> <your-private-repo>/busybox

Adding signer "<your-name>" to <your-private-repo>/busybox...
Enter passphrase for repository key with ID b0014f8: 
Successfully added signer: <your-name> to <your-private-repo>/busybox
```

> If you create a new signature for the image now, you'll be ask for your delegation key instead of the targets key, therefore creating a signature using the delegation.

Reviewer: ```suggestion
If you create a new signature for the image, you'll be asked for your delegation key instead of the targets key, therefore creating a signature using the delegation.
```

```bash
> DOCKER_CONTENT_TRUST=1 docker push <your-private-repo>/busybox:stable
```

> To tell Connaisseur that this signer/delegation is required to be present on an images' signature, the `with.delegations` list inside an image policy rule can be used. Simply add the signers name to the list. You can also add multiple signer names to the list, but be aware that Connaisseur currently only supports an AND conjunction, meaning every signer listed must have signed the matching image.

Reviewer: Signer is not required on the image signature, but rather the signer's signature is required on the image's trust data

> To tell Connaisseur that this signer/delegation is required to be present on an images' signature, the `with.delegations` list inside an image policy rule can be used. Simply add the signers name to the list. You can also add multiple signer names to the list, but be aware that Connaisseur currently only supports an AND conjunction, meaning every signer listed must have signed the matching image.

Reviewer: ```suggestion
To tell Connaisseur that this signer/delegation is required to be present on an image's signature, the `with.delegations` list inside an image policy rule can be used. Simply add the signer's name to the list. You can also add multiple signer names to the list, but be aware that Connaisseur currently only supports an AND conjunction, meaning every signer listed must have signed the matching image.
```
I think thats the way. Not sure though

```

Connaisseur now rejects all images from the given repository that have not been signed based on the provided public key.
A quick guide for installation and testing is available in [getting started](../getting_started.md#deploy-connaisseur).
It also provides a full step-by-step guide.

### Understanding validation

Using the simple pre-configuration shipped with Connaisseur, it is possible to test validation by deploying some pods:

```bash
kubectl run test-signed --image=docker.io/securesystemsengineering/testimage:signed
> pod/test-signed created

kubectl run test-unsigned --image=docker.io/securesystemsengineering/testimage:unsigned
> Error from server: admission webhook "connaisseur-svc.connaisseur.svc" denied the request: Unable to find signed digest for image docker.io/securesystemsengineering/testimage:unsigned.
# or in case of a signature with a different key
> Error from server: admission webhook "connaisseur-svc.connaisseur.svc" denied the request: Failed to verify signature of trust data root.
```

How does Connaisseur validate these requests and convert the images with tags to digests?

> What happens in the background is that Connaisseur looks up trust data of the image in the `root`, `snapshot`, `timestamp` and `targets` files (in json format) by querying the API of the Notary server.

Reviewer: ```suggestion
What happens in the background is that Connaisseur looks up trust data of the image in the `root`, `snapshot`, `timestamp` and `targets` files (in JSON format) by querying the API of the Notary server.
```

Trust data syntax is validated against their known schemas and the files' signatures are validated against their respective public keys.
The pinned root key is used for the `root.json` file that in turn contains the other keys which can then be trusted for validation of the remaining trust data (`snapshot.json`, `timestamp.json`, `targets.json`).
Furthermore, Connaisseur gathers trust data of potential delegations linked in the `targets` file which can then be used to [enforce delegations](#enforcing-delegations).

> At this point, Connaisseur is left with a set of potentially disjoint or overlapping sets of trust data.

Reviewer 2: I think this sentence is hard to understand. Before, we spoke about the trust data and now it's sets of trust data and it's unclear how these sets are derived.

Reviewer: Maybe something along
```
At this point, Connaisseur is left with a set of potentially disjoint or overlapping sets of trust data (i.e. signatures on digests).
Connaisseur filters the trust data for signed digests that actually relate to the image under validation.
In case exactly one trusted digest remains, Connaisseur modifies the admission request and admits it.
Otherwise, admission is rejected (this could happen if Alice signs image:tag linking to image:shaabc and Bob signs image:tag linking to image:shadef.
```
?

Author: proposal

Connaisseur filters the trust data for signed digests that actually relate to the image under validation.
In case exactly one trusted digest remains, Connaisseur modifies the admission request and admits it.
Otherwise, admission is rejected.

While it is obvious to reject an image that does not exhibit a trusted digest, there is the special case of multiple trusted digests.
This only occurs in some edge cases, but at this point Connaisseur cannot identify the right digest anymore and consequently has to reject.

For more information on TUF roles, please refer to [TUF's documentation](https://theupdateframework.io/security/) or checkout [this introductory presentation](https://berlin-crypto.github.io/event/dockerimagesignatures.html) on how the trust data formats work and are validated by Connaisseur.

## Configuration options

`.validators[*]` in `helm/values.yaml` supports the following keys for Notary (V1) (refer to [basics](../basics.md#validators) for more information on default keys):

| key | default | required | description |
| - | - | - | - |

> | `name` | - | :heavy_check_mark: | see [basics](../basics.md#validators) |

Reviewer: ```suggestion
| `name` | - | :heavy_check_mark: | See [basics](../basics.md#validators) |
```

| `type` | - | :heavy_check_mark: | `notaryv1`; the validator type must be set to `notaryv1`. |
| `host` | - | :heavy_check_mark: | URL of the Notary instance, in which the signatures reside, e.g. `notary.docker.io`. |

> | `trust_roots[*].name` | - | :heavy_check_mark: | see [basics](../basics.md#validators) |

Reviewer: ```suggestion
| `trust_roots[*].name` | - | :heavy_check_mark: | See [basics](../basics.md#validators) |
```

> | `trust_roots[*].key` | - | :heavy_check_mark: | see [basics](../basics.md#validators). ECDSA root public key. |

Reviewer: ```suggestion
| `trust_roots[*].key` | - | :heavy_check_mark: | See [basics](../basics.md#validators). ECDSA root public key. |
```

> Adding signer "<key-name>" to <your-repo>/busybox...
> Enter passphrase for repository key with ID b0014f8: 
> Successfully added signer: <key-name> to <your-repo>/busybox
```

If you create a new signature for the image, you'll be asked for your delegation key instead of the targets key, therefore creating a signature using the delegation.

```bash
> DOCKER_CONTENT_TRUST=1 docker push <your-repo>/busybox:stable
```

Without further configuration, Connaisseur will accept all delegation signatures for an image that can ultimately be validated against the root public key.
Connaisseur can enforce a certain signer/delegation (or multiple) for an image's signature via the `with.delegations` list inside an image policy rule.
Simply add the signer's name to the list.
You can also add multiple signer names to the list in which case Connaisseur will enforce that *all* delegations must have signed a matching image.

```yaml
policy:
- pattern: "<your-repo>/busybox:*"
  with:
    delegations:
    - <key-name>
    - <other-key-name>
```

The delegation feature can be useful in complex organisations where certain people may be required to sign specific critical images.

> Another use-case is to sign an image with delegation keys in various stages of your CI and enforce that certain checks were passed, i.e. enforcing the signature of your linter, your security scanner and your software lisence compliance check.

Reviewer: ```suggestion
Another use-case is to sign an image with delegation keys in various stages of your CI and enforce that certain checks were passed, i.e. enforcing the signature of your linter, your security scanner and your software license compliance check.
```
or licence for british english, dunno which we use actually 😅
